### PR TITLE
windows: Fix JSON schema validation

### LINKED ISSUE #13394: windows: The json-language-server does not support JSON schema validation
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Open `Zed`, open `settings.json`. On macOS, there is some kind of auto-completion supports based on json-schema. But no JSON schema validation on windows.

This is a issue of [vscode-json-languageserver](https://www.npmjs.com/package/vscode-json-languageserver) that `Zed` depends on. `vsocde-json-languageserver` relies on an older version of [vscode-json-languageservice](https://github.com/microsoft/vscode-json-languageservice), which has issues with JSON schema validation on Windows. See https://github.com/microsoft/vscode/issues/176296 and https://github.com/microsoft/vscode-json-languageservice/pull/177.

This issue was resolved in version `vscode-json-languageservice v5.3.2`, but since the `vscode-json-languageserver` hasn't been updated for a long time, the problem still persists.

### Environment

windows 11

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by JunkuiZhang**
This PR needs suggestions, especially from the Zed team. As I mentioned in a previous issue #13394 , the `vscode-json-languageserver` that Zed originally relied on has some issues with JSON schema validation on Windows, and it hasn't been updated for a long time. This PR uses the more frequently updated `vscode-langservers-extracted`, which resolves this issue.

Currently, `vscode-langservers-extracted` includes not only the JSON LSP server but also LSP servers for other languages. I think we might need a package specifically for the JSON LSP server, such as something like `vscode-json-langserver-extracted`, or we could consider using the LSP servers for other languages from this package as well. 

And, there are some issues with installing `vscode-langservers-extracted` on Windows, causing the `postinstall` script to fail. However, this does not seem to affect any functionality. Therefore, I think the best solution is for the Zed team to maintain a package like `vscode-json-langserver-extracted` or something else. This way, we can update it promptly and address the installation issues on Windows.

Any suggestions or advices are welcome.

#### JSON vaildation on Winodws

https://github.com/zed-industries/zed/assets/14981363/8cd7ff54-28ec-4601-b2e5-183e2fae2051

Closes #13394 

Release Notes:

- Fixed JSON schema validation issue on Windows.(#13394 )

### CODE REVIEW

File: crates/languages/src/json.rs [hunk 1]
```diff
@@ -156,7 +157,8 @@ impl LspAdapter for JsonLspAdapter {
         if should_install_language_server {
             self.node
                 .npm_install_packages(&container_dir, &[(package_name, latest_version.as_str())])
-                .await?;
+                .await
+                .log_err();
```
Comment (JunkuiZhang) [inline crates/languages/src/json.rs:161]: On windows, this will fail, the node log:
```txt
230 info run core-js@3.37.1 postinstall node_modules/core-js node -e "try{require('./postinstall')}catch(e){}"
231 info run core-js@3.37.1 postinstall { code: undefined, signal: undefined }
232 verbose stack TypeError [ERR_INVALID_ARG_TYPE]: The "file" argument must be of type string. Received undefined
232 verbose stack     at new NodeError (node:internal/errors:399:5)
232 verbose stack     at validateString (node:internal/validators:163:11)
232 verbose stack     at normalizeSpawnArguments (node:child_process:545:3)
232 verbose stack     at spawn (node:child_process:750:13)
232 verbose stack     at C:\Users\36477\AppData\Local\Zed\node\node-v18.15.0-win-x64\node_modules\npm\node_modules\@npmcli\promise-spawn\lib\index.js:18:12
232 verbose stack     at new Promise (<anonymous>)
232 verbose stack     at promiseSpawn (C:\Users\36477\AppData\Local\Zed\node\node-v18.15.0-win-x64\node_modules\npm\node_modules\@npmcli\promise-spawn\lib\index.js:17:13)
232 verbose stack     at spawnWithShell (C:\Users\36477\AppData\Local\Zed\node\node-v18.15.0-win-x64\node_modules\npm\node_modules\@npmcli\promise-spawn\lib\index.js:123:10)
232 verbose stack     at promiseSpawn (C:\Users\36477\AppData\Local\Zed\node\node-v18.15.0-win-x64\node_modules\npm\node_modules\@npmcli\promise-spawn\lib\index.js:12:12)
232 verbose stack     at runScriptPkg (C:\Users\36477\AppData\Local\Zed\node\node-v18.15.0-win-x64\node_modules\npm\node_modules\@npmcli\run-script\lib\run-script-pkg.js:79:13)
233 verbose cwd C:\Users\36477\AppData\Local\Zed\languages\json-language-server
234 verbose Windows_NT 10.0.22631
235 verbose node v18.15.0
236 verbose npm  v9.5.0
237 error code ERR_INVALID_ARG_TYPE
238 error The "file" argument must be of type string. Received undefined
239 verbose exit 1

```

**Comment by osiewicz** [comment]
For what it's worth, we already maintain our own fork of HTML language server, so I think it would make sense to maintain a fork of JSON language server as well. 
Btw. it'd be lovely if you could open an issue for this postinstall failure with HTML language server.

**Review by osiewicz — APPROVED** [review]
LGTM

File: crates/languages/src/json.rs [hunk 2]
```diff
@@ -240,7 +242,19 @@ async fn get_cached_server_binary(
...
         .unwrap()
+        .display()
+        .to_string();
+    #[cfg(target_os = "windows")]
+    return path
+        .strip_prefix(path.parent().unwrap().parent().unwrap())
+        .unwrap()
+        .display()
+        .to_string()
+        .replace("\\", "/");
```
Comment (JunkuiZhang) [inline crates/languages/src/json.rs:259]: On Windows, this function results in `"Zed\settings.json"`, which is not currently supported by `fileMatch`. Therefore, it is converted to `"Zed/settings.json"`.

**Comment by JunkuiZhang** [comment]
Thank you very much for your response. This PR is somewhat of a workaround; is it acceptable?

**Comment by osiewicz** [comment]
Yes, I think it's acceptable as is. We can look into migrating to our own fork of JSON language server eventually, but if that fixes the issue short-term, that's fine by me.

File: crates/languages/src/json.rs (hunk 2 above)
Comment (osiewicz) [reply crates/languages/src/json.rs:259]: Tbh we can just skip the `cfg` dance and use `.replace('\\', "/")` on both platforms. This code isn't really hot.

Comment (JunkuiZhang) [reply crates/languages/src/json.rs]: Thanks for your suggestion. I've removed the `cfg`s.

**Comment by osiewicz** [comment]
Do you want to make any other changes? This looks mergeable to me as is, so unless you want to do anything else, I'll just go ahead and merge.

**Comment by JunkuiZhang** [comment]
No. Thanks for your help!

**Comment by mrnugget** [comment]
This PR broke schema validation for me on macOS. Autocomplete still works, but diagnostics are missing, which is very important for our settings files.

See here: the `00000` is not highlighted as an error. The new language server here just sends `[]` back as diagnostics. If I revert this PR locally I get diagnostics back.

screenshot-2024-07-03-09 11 58@2x


I think we should revert this for now, since Windows isn't officially supported yet and without schema validation of settings/keymaps, many users run into issues. We should also revert this in the Preview branch, since this change is already in there.

**Comment by osiewicz** [comment]
Since we've started using a more recent version of language server, we've ran into issues with getting empty diagnostics. I pinned it down to a difference between the old version of a server and the new one. Windows has nothing to do with it.
TL;DR: we get empty diagnostics due to [this line](https://github.com/microsoft/vscode-json-languageservice/blame/main/src/services/jsonValidation.ts#L36), which in turn is triggered in more recent server by [this more recent line that's included in the new version of a JSON server](https://github.com/microsoft/vscode/blob/main/extensions/json-language-features/server/src/jsonServer.ts#L240)
Obviously, this being JS, this has to have a bunch of papercuts :) In this case the intention was to fall back to `true` value if `validate` key is not set at all, but instead.. ah, damn, just see for yourself:
```
  let settings = {json: {}}

let validateEnabled = !!settings.json?.validate?.enable;
  console.log(validateEnabled)

  let settings2 = {json: {validate: {enable: true}}}

let validateEnabled2 = !!settings2.json?.validate?.enable;
  console.log(validateEnabled2)
```
Setting `validate` explicitly in our initialization options for the server fixes it. I'll bump Preview and nightly with it.

**Comment by JunkuiZhang** [comment]
@osiewicz Thank you for diving into this issue and fix it. Sorry for introducing this bug.

**Comment by osiewicz** [comment]
No problem, that's not your fault. =)
